docs/architecture-decisions/0009-support-callbacks-after-processing-messages.md

```diff
@@ -0,0 +1,31 @@
+# 9. Support callbacks after processing messages
+
+Date: 2022-05-04
+
+## Status
+
+Accepted
+
+## Context
+
+DSS currently requires manual intervention. This was accepted during the initial build-out, but now that it has stabilized in production it is time to provide the ability for applications to move to full automation.
+
+Triggering DSS to run after a batch has been submitted is already possible using direct AWS calls so DSS will not concern itself with how that works. For instance, ETD knows when it has finished populating the SQS queue and can easily add a call to start the fargate task directly that time if permissions are updated in AWS to allow that. Wiley should be able to do this as well. So starting processing on the incoming queue will be solved in the upstream applications.
+
+To allow the upstream applications to know when to process their output queues, we'll need to make a change to both the incoming message specification and the application logic. This will be optional. Upstream applications can choose to continue to manually trigger DSS and the processing of output queues if they have reason to prefer that.
+
+We'll update the incoming message spec to allow for including a callback URL in each message. Upstream applications can provide a URL that will alert them to process the output queue. As many uses cases will be to only receive one callback for a batch, DSS will keep track of all callback URLs during a processing batch and when it receives no additional SQS messages to process, it will add a step to call each _unique_ URL it encountered. The ETD application will listen on a route for this and then kick off it's existing output messaging logic. Wiley could consider creating a lambda that is triggered by a URL and providing that to DSS and then put any processing logic for the output queue in that lambda.
+
+## Decision
+
+We will allow an optional Callback URL in the input message to DSS.
+
+Upon finishing processing incoming messages, DSS will call all unique callback URLs it encountered.
```

Comment:
Is there a scenario where two upstream applications posting an overlapping batch of messages could cause a problem here? Let's say:

1. ETD and Wiley both start sending messages to the queue at roughly the same time.
2. Wiley finishes sending messages first and start DSS processing. ETD is still sending messages in its batch.
3. DSS starts processing messages. Its callback list contains both the Wiley and ETD callback URLs.
4. For whatever reasons around lag time or just a job that takes a while, DSS gets through all the messages and ETD is still sending its batch, but there's a longer than 20 second lag between messages sent to the queue from ETD.
5. DSS gets an empty response after long polling and calls the Wiley and ETD callback URLs.
6. ETD keeps sending messages. When it finishes the batch, it starts DSS again.
7. DSS processes the remaining ETD messages from the batch and calls the callback URL for those. This _may_ happen while the first post-callback process is still happening in ETD (another overlapping process scenario).

I think this is both super unlikely and will be fine if it happens, but it does mean the post-callback output process used by the upstream apps should be able to handle multiple callback jobs for a single batch and ensure those jobs don't conflict with each other, just in case.

Am I missing anything? I know this is super unlikely but I don't want us to end up with borked data in ETD because of some bizarre outlier scenario.

Reply:
I think you are both correct that this is unlikely and still needs to be accounted for. ETD/Wiley would need to be able to know what to do if they get a message to start the output queue processing. They might choose to check if they are already processing the queue (and thus ignore the second message) or send an alert that something weird happened or... something else.

We'll similarly want to ensure the DSS fargate task checks to see if it is already running before spinning up an additional task to avoid sending more data to dspace than it can handle.